### PR TITLE
New Algorithm: CalculateEfficiency v2

### PR DESCRIPTION
**Pull request by quantumsteve**
**Description of work.**

This adds the CalculateSensitivity algorithm used by the the EQSANS and HFIR SANS reduction.

Written by @ricleal, with some linting done by me.

**To test:**

Review code & unit test builds.

Part of #25095.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gvardany** [comment]
I think we run into the same story; can you explain why not to use the existing [CalculateEfficiency ](https://docs.mantidproject.org/nightly/algorithms/CalculateEfficiency-v1.html)?

**Comment by ricleal** [comment]
The new `CalculateSensitivity` algorithm is different from the old `CalculateEfficiency`.

Here my notes:

Description:
===========

This ticket only calculates the sensitivity => Outputs a nexus file!

Delete / Replace:
=================

* [ ]  `SANSSensitivityCorrection` (Workflow algorithm).

This is used to calculate the sensitivity and to apply it (?)

In addition to that, it does:
- dark current subtraction
- beam centre (needed for solid angle)
- solid angle correction
- monitor normlaization

Modify (creat enew?):
=======
* `CalculateEfficiency`:
    * [x]  Remove the mask properties + adapt the code. New algorthm `CalculateSensitivity`. 
 Branch: `ORNL_SANS_Sensitivity_Calculation` in mantid.

GPSANS
======

To see how the patch work see: https://github.com/ricleal/HFIR/blob/master/scripts/sensitivity_20170818.py

## Full Algorithm:

### Sensitivity calculation

```
for every flood offset (0, 200 and 400):
  - Apply any masks to the flood

  - subtract dark current from the flood (if a dark current is given)

  - find the beam center from the flood unless it's specified a center (x,y)

  - calculate and correct the flood for solid angle

  - if flood transmission is available, do transmission correction to the flood

  - calculate_sensitivity algorithm for the flood

  - mask flood beamstop and keep the resulting workspace

 

merge/average the workspaces for every offset (0, 200 and 400)

result S(x,y) = a single workspace from the 3 offsets
```

### Application of the sensitivity 

Application of the sensitivity correction is a straightforward division of the data: the data are divided by the sensitivity S(x,y)

# Algorithm for Biosans & EQSANS

**A pixel should have:**
1. a value
2. be masked 
3. or "marked". 

In the documentation they use "inf" to mark a pixel. In Mantid when you have a pixel with "inf" counts you cannot use the instrument visualisation (that it's need for additional masking)...
- Use mantid EMPTY_INT rather then "inf".

**Procedure:**
- Calculate the average and normalize
- Apply a bad pixel threshold
  - If it lies within the range, it is retained. 
  - If not, the value is set to EMPTY_INT to flag it as locally masked, but it is treated differently than masked pixels are during the remainder of the sensitivity file preparation.
- Patch the sensitivity (interpolation of a mask)
  - If a pixel was marked as non-performing (i.e. EMPTY_INT), it is ignored during the calculations performed during the patching and its value is not overwritten.

**Comment by gvardany** [comment]
It seems that the main difference is the EMPTY_INT instead of inf, and the treatment of the masks. Otherwise the actual calculation is not different. If you insist that the differences can not be easily incorporated to `CalculateEfficiency`, then it should be at least `CalculateEfficiency-v2`. I can not tolerate having two distinct but alarmingly similar algorithms; this creates more confusion, and there is enough of it already. Also I do not see the point of making this instrument specific; this algorithm merely takes the corrected flood data and generates the sensitivity map, which is a generic operation. The definition of sensitivity is the same for everyone, so my point is that we have to make effort in consolidating this down to one algorithm for all the SANS instruments in mantid.

**Comment by quantumsteve** [comment]
Closing to address properly at a later point in time.

**Comment by peterfpeterson** [comment]
Getting this in shape based on @gvardany's comments.

**Review by jmborr — DISMISSED** [review]
Everything looks good. The tests are very thorough :+1:

File: Framework/Algorithms/src/CalculateEfficiency2.cpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                      PropertyNames::INPUT_WORKSPACE, "", Direction::Input),
+                  "The workspace containing the flood data");
+  declareProperty(
+      std::make_unique<WorkspaceProperty<>>(PropertyNames::OUTPUT_WORKSPACE, "",
+                                            Direction::Output),
+      "The name of the workspace to be created as the output of the algorithm");
+
+  auto positiveDouble = boost::make_shared<BoundedValidator<double>>();
+  positiveDouble->setLower(0);
+  declareProperty(PropertyNames::MIN_THRESHOLD, 0.0, positiveDouble,
```
[the remark this answers is not in the document]
Comment (peterfpeterson) [reply Framework/Algorithms/src/CalculateEfficiency2.cpp]: This seems like a new issue (#26059) that requires a bit of research.

[the remark this answers is not in the document]
Comment (jmborr) [reply Framework/Algorithms/src/CalculateEfficiency2.cpp]: Created `sans-backend` issue #[98](https://code.ornl.gov/sns-hfir-scse/sans/sans-backend/issues/98)

**Review by jmborr — APPROVED** [review]

**Pull request merged**